src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1594,6 +1596,101 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             goto Finish;
         }
 
+        private static unsafe nuint NarrowUtf16ToAscii_AdvSimd(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+        {
+            // This meethod follows similar logic as the NarrowUtf16ToAscii_Sse2 method above.
```

Comment:
Can the cross-platform Vector128 methods be used to replace what's there instead of duplicating code?

Reply:
Tried using Vector128 methods wherever possible such as, load/stores. However, couldn't find suitable method for operations such as, pairwise maximum and unzip.
Felt that combining both x86 and AArch64 versions may reduce readability with more platform specific code.

Reply:
> couldn't find suitable method

@tannergooding?

> Felt that combining both x86 and AArch64 versions may reduce readability with more platform specific code.

We really want to use the new Vector128 methods for such things whenever possible.

Reply:
> Felt that combining both x86 and AArch64 versions may reduce readability with more platform specific code.

Yeah, we really want to have a single path where possible. Small bits of functionality that differ, like a `pairwise max` can be extracted to their own helper function so that the main path remains "simple/readable".

Notably x86/x64 doesn't have a pairwise max, you at best get "shuffle/unzip + max". For `unzip`, you just want to use `Shuffle`. If it takes a single vector, that API is already available, if it takes two different vectors then you'll need an intermediate helper method as those APIs are a WIP on my end.

Reply:
> Yeah, we really want to have a single path where possible.

Would it make sense to do that as a separate follow on patch? Net 7.0 deadlines are approaching fast, and we'd like to get in this, GetIndexOfFirstNonAsciiChar, and the two system.buffers.base64 routines (first patch for that is imminent after the UK jubilee weekend).
Keeping them separate means any issues are limited to getting the best/fastest code for Arm64, and ensuring use of vector128 api where possible.
The single path versions have a lot more scope for design issues - what if that is non-optimal on one target? should helper functions go into the vector128 api? Plus there is all the additional testing (and neither of us have any x86 machines setup yet). None of those are a major hurdle, but it feels best to push that to post the 7.0 release. On the face of it, NarrowUtf16ToAscii seems simple enough to merge (so it may still be possible to slip it in in time), but the others I'm not so sure about (the base64 routines already have two X86 versions using the old APIs).

Reply:
The new cross-platform APIs are intended to be the go-to mechanism for this kind of thing, only falling back to the lower-level intrinsics when the operation can't be expressed with those APIs. We have an open issue to convert most of the existing intrinsics usage to the new APIs, and we need to stop creating debt. We also need to validate the new APIs, both from an expressibility/usability y perspective and from a performance perspective. They're new in. NET 7 and we need such validation in order to ship them. I would rather see fewer APIs vectorized but using the new APIs than more vectorized using the intrinsics, but the intent is that using the new APIs should be overall easier, and breed more consistency and maintainability. Let's focus our efforts there.

Reply:
This code is in a halfway place - it's using the new APIs (where possible), but is still split out as a completely different function. That's showing the API is fine from a performance view, but agreed, it's not showing expressibility/usability.

Reply:
Instead of forking the entire logic in a separate method `NarrowUtf16ToAscii_AdvSimd`, it will be good to have all the logic in one method (rename `NarrowUtf16ToAscii_SSE` with `NarrowUtf16ToAscii_Intrinsified` or something) and then guard them with `AdvSimd.IsSupported` vs. `SSE2.IsSupported`. We have done that in past, eg.

https://github.com/dotnet/runtime/blob/1b4a9c47ba8e4ad53562e003240527b738b0e7d1/src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs#L391-L406

JIT will make sure that the condition is `AdvSimd.IsSupported` is const-prop and hence won't see a branch plus it will have an advantage of just converting one method (instead of two) to cross-platform Vector APIs in future.

Reply:
Added `NarrowUtf16ToAscii_Intrinsified` method that combines SSE and ASIMD logic.